internal/mesh/internal/controllers/sidecarproxy/builder/destination_builder.go

```diff
@@ -271,8 +278,56 @@ func (b *Builder) buildDestination(
 		)
 		clusterName := fmt.Sprintf("%s.%s", portName, sni)
 
-		b.addCluster(clusterName, sni, portName, details.IdentityRefs)
-		b.addEndpointsRef(clusterName, details.ServiceEndpointsId, details.MeshPort)
+		if details.FailoverConfig != nil {
+			failover := details.FailoverConfig
+			// TODO(v2): handle other forms of failover (regions/locality/etc)
+
+			var endpointGroups []*pbproxystate.EndpointGroup
+			for i, dest := range failover.Destinations {
+				if dest.BackendTarget == types.NullRouteBackend {
+					panic("not possible")
+				}
+				destDetails, ok := targets[dest.BackendTarget]
+				if !ok {
+					panic("not possible")
+				}
+
+				destDC := defaultDC(destDetails.BackendRef.Datacenter)
+				destPortName := destDetails.BackendRef.Port
+
+				destSNI := DestinationSNI(
+					destDetails.BackendRef.Ref,
+					destDC,
+					b.trustDomain,
+				)
+				destClusterName := fmt.Sprintf("%s%d~%s", xdscommon.FailoverClusterNamePrefix, i, clusterName)
+
+				egDest := b.newClusterEndpointGroup(destClusterName, destSNI, destPortName, destDetails.IdentityRefs)
```

Comment:
Each inner endpoint group gets its name assigned here, so that we have the name available to create an endpoints ref (for EDS)